File: Homework 1/swapMethods.java
import java.io.*;
import java.util.Random;
import java.lang.reflect.Method;

public class swapMethods {

// Temporary Variable swap
    static void swapByTemp(int a, int b) {
        System.out.println("x = "+ a);
        System.out.println("y = "+ b);
        int temp = a;
        a = b;
        b = temp;
        System.out.println("x = "+ a);
        System.out.println("y = "+ b);
    }

// Bitwise swap
    static void swapByXOR(int a, int b) {
        System.out.println("x = "+ a);
        System.out.println("y = "+ b);
        a = a ^ b;
        b = a ^ b;
        a = a ^ b;
        System.out.println("x = "+ a);
        System.out.println("y = "+ b);
    }

// Add/Sub swap
    static void swapByAddSub(int a, int b) {
        System.out.println("x = "+ a);
        System.out.println("y = "+ b);
        a = a + b;
        b = a - b;
        a = a - b;
        System.out.println("x = "+ a);
        System.out.println("y = "+ b);
    }

// Mult/Div swap
    static void swapByMultDiv(int a, int b) {
        System.out.println("x = "+ a);
        System.out.println("y = "+ b);
        a = a * b;
        b = a / b;
        a = a / b;
        System.out.println("x = "+ a);
        System.out.println("y = "+ b);
    }
// Function for calculating averages
    static void takeAvgTime(long[] list) {
        float avgTime = 0;
        for(int i = 0; i < list.length; i++){
            avgTime += list[i];
            }
        avgTime = avgTime / 10000;
        System.out.println(avgTime);
    }

// Main Driver Function
    public static void main(String a[]) {
        Random rand = new Random();
		// Sets necessary variables and lists for functiions
        int counter = 0;
        long[] tempList = new long[10000];
        long[] XORList = new long[10000];
        long[] AddSubList = new long[10000];
        long[] MultDivList = new long[10000];
        long beforeTime = 0;
        long afterTime = 0;
        long oneAvg;
		// Loop for conducting the swapping process
        while (counter < 10000) {
			// Sets random numbers
            int y = rand.nextInt(1, 10000);
            int x = rand.nextInt(1, 10000) + y;
			// Sets time before and after the running of each method and stores time difference to the given list
            beforeTime = (System.currentTimeMillis() * 1000);
            swapByTemp(x, y);
            afterTime = (System.currentTimeMillis() * 1000);
            oneAvg = afterTime - beforeTime;
            tempList[counter] = oneAvg;
            beforeTime = (System.currentTimeMillis() * 1000);
            swapByXOR(x, y);
            afterTime = (System.currentTimeMillis() * 1000);
            oneAvg = afterTime - beforeTime;
            XORList[counter] = oneAvg;
            beforeTime = (System.currentTimeMillis() * 1000);
            swapByAddSub(x, y);
            afterTime = (System.currentTimeMillis() * 1000);
            oneAvg = afterTime - beforeTime;
            AddSubList[counter] = oneAvg;
            beforeTime = (System.currentTimeMillis() * 1000);
            swapByMultDiv(x, y);
            afterTime = (System.currentTimeMillis() * 1000);
            oneAvg = afterTime - beforeTime;
            MultDivList[counter] = oneAvg;
            counter++;
            }
		// Uses list to compute averages and displays them
        takeAvgTime(tempList);
        takeAvgTime(XORList);
        takeAvgTime(AddSubList);
        takeAvgTime(MultDivList);
    }
}

File: Homework 1/swapMethods.py
import random
from time import time

## Temporary variable swap
def swapByTemp(a, b):
    print("X is : ", int(a))
    print("Y is : ", int(b))
    temp = a
    a = b
    b = temp
    print("X is : ", int(a))
    print("Y is : ", int(b))

##Bitwise swap
def swapByXOR(a, b):
    print("X is : ", int(a))
    print("Y is : ", int(b))
    a = a ^ b
    b = a ^ b
    a = a ^ b
    print("X is : ", int(a))
    print("Y is : ", int(b))

## Add/Sub swap
def swapByAddSub(a, b):
    print("X is : ", int(a))
    print("Y is : ", int(b))
    a = a + b
    b = a - b
    a = a - b
    print("X is : ", int(a))
    print("Y is : ", int(b))
    
## Mult/Div swap
def swapByMultDiv(a, b):
    print("X is : ", int(a))
    print("Y is : ", int(b))
    a = a * b
    b = a / b
    a = a / b
    print("X is : ", int(a))
    print("Y is : ", int(b))

## Function to conduct one time average given an output list and swap function
def takeOneTime(listEx, func):
    ## Generates Random Numbers
    y = random.randrange(1, 10000)
    x = random.randrange(1, 10000) +y
    ## Calculates before and after times, uses swap method, and computes the time
    beforeTime = int(time() * 1000000)
    func(x, y)
    afterTime = int(time() * 1000000)
    oneAvg = afterTime - beforeTime
    ## Time taken is stored to list
    listEx.append(oneAvg)
    return listEx

## Function that takes lists from takeOneTime and computes averages for each method
def takeAvgTime(outList):
    avgTime = 0
    for i in outList:
        avgTime += i
    avgTime = avgTime / 10000
    return avgTime

## Main Driver Function
def main():
## Initializes necessary variables
    counter = 0
    tempList = []
    XORList = []
    AddSubList = []
    MultDivList = []
    avgTemp = 0
    avgXOR = 0
    avgAddSub = 0
    avgMultDiv = 0
    ## Loop for conducting swap and instantial times
    while counter <= 10000:
        takeOneTime(tempList, swapByTemp)
        takeOneTime(XORList, swapByXOR)
        takeOneTime(AddSubList, swapByAddSub)
        takeOneTime(MultDivList, swapByMultDiv)
        counter += 1
       ## Computes averages and prints them out
    avgTemp = takeAvgTime(tempList)
    avgXOR = takeAvgTime(XORList)
    avgAddSub = takeAvgTime(AddSubList)
    avgMultDiv = takeAvgTime(MultDivList)
    print("Average time for Temp: ", avgTemp, "us")
    print("Average time for XOR: ", avgXOR, "us")
    print("Average time for AddSub: ", avgAddSub, "us")
    print("Average time for MultDiv: ", avgMultDiv, "us")

main()
